staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller_debug.go

```diff
@@ -0,0 +1,199 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/server/mux"
+)
+
+func (cfgCtl *configController) Install(c *mux.PathRecorderMux) {
+	// TODO(yue9944882): handle "Accept" header properly
+	// debugging dumps a CSV content for three levels of granularity
+	// 1. row per priority-level
+	c.UnlistedHandleFunc("/debug/api_priority_and_fairness/dump_priority_levels", cfgCtl.dumpPriorityLevels)
+	// 2. row per queue
+	c.UnlistedHandleFunc("/debug/api_priority_and_fairness/dump_queues", cfgCtl.dumpQueues)
+	// 3. row per request
+	c.UnlistedHandleFunc("/debug/api_priority_and_fairness/dump_requests", cfgCtl.dumpRequests)
+}
+
+func (cfgCtl *configController) dumpPriorityLevels(w http.ResponseWriter, r *http.Request) {
+	cfgCtl.lock.Lock()
+	defer cfgCtl.lock.Unlock()
+	tabWriter := tabwriter.NewWriter(w, 8, 0, 1, ' ', 0)
+	columnHeaders := []string{
+		"PriorityLevelName",
+		"ActiveQueues",
+		"IsIdle",
+		"WaitingRequests",
+		"ExecutingRequests",
+	}
+	tabPrint(tabWriter, rowForHeaders(columnHeaders))
+	for _, plState := range cfgCtl.priorityLevelStates {
+		if plState.queues == nil {
+			continue
+		}
+		queueSetDigest := plState.queues.Dump()
+		activeQueueNum := 0
+		for _, q := range queueSetDigest.Queues {
+			if len(q.Requests) > 0 {
+				activeQueueNum++
+			}
+		}
+
+		tabPrint(tabWriter, rowForPriorityLevel(
+			plState.pl.Name,
+			activeQueueNum,
+			plState.queues.IsIdle(),
+			queueSetDigest.Waiting,
+			queueSetDigest.Executing,
+		))
+	}
+	runtime.HandleError(tabWriter.Flush())
+}
+
+func (cfgCtl *configController) dumpQueues(w http.ResponseWriter, r *http.Request) {
+	cfgCtl.lock.Lock()
+	defer cfgCtl.lock.Unlock()
+	tabWriter := tabwriter.NewWriter(w, 8, 0, 1, ' ', 0)
+	columnHeaders := []string{
+		"PriorityLevelName",
+		"Index",
+		"PendingRequests",
+		"ExecutingRequests",
+		"VirtualStart",
+	}
+	tabPrint(tabWriter, rowForHeaders(columnHeaders))
+	for _, plState := range cfgCtl.priorityLevelStates {
+		if plState.queues == nil {
+			tabPrint(tabWriter, row(
+				plState.pl.Name,
+				"<none>",
+				"<none>",
+				"<none>",
+				"<none>",
+			))
+			continue
+		}
+		queueSetDigest := plState.queues.Dump()
+		for i, q := range queueSetDigest.Queues {
+			tabPrint(tabWriter, rowForQueue(
+				plState.pl.Name,
+				i,
+				len(q.Requests),
+				q.ExecutingRequests,
+				q.VirtualStart,
+			))
+		}
+	}
+	runtime.HandleError(tabWriter.Flush())
+}
+
+func (cfgCtl *configController) dumpRequests(w http.ResponseWriter, r *http.Request) {
+	cfgCtl.lock.Lock()
+	defer cfgCtl.lock.Unlock()
+	tabWriter := tabwriter.NewWriter(w, 8, 0, 1, ' ', 0)
+	columnHeaders := []string{
+		"PriorityLevelName",
+		"FlowSchemaName",
+		"QueueIndex",
+		"RequestIndexInQueue",
+		"FlowDistingsherType",
```

Comment:
I still do not favor having a column for FlowDistingsherType, because it is denormalized data (every request of a given FlowSchema has the same value in this column).  This is not a show-stopper for me.

Reply:
removed `FlowDistingsherType`, and also removed `StartTime` as it will be always be zero for requests in queue